test/parallel/test-tls-connect-allow-half-open-option.js

```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+const common = require('../common');
+
+// This test verifies that `tls.connect()` honors the `allowHalfOpen` option.
+
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const assert = require('assert');
+const fixtures = require('../common/fixtures');
+const tls = require('tls');
+
+{
+  const socket = tls.connect({ lookup() {} });
+  assert.strictEqual(socket.allowHalfOpen, false);
+}
+
+{
+  const socket = tls.connect({ allowHalfOpen: false, lookup() {} });
+  assert.strictEqual(socket.allowHalfOpen, false);
+}
+
+
+const server = tls.createServer({
+  key: fixtures.readKey('agent1-key.pem'),
+  cert: fixtures.readKey('agent1-cert.pem'),
+}, common.mustCall((socket) => {
+  server.close();
+
+  socket.setEncoding('utf8');
+  socket.on('data', common.mustCall((chunk) => {
+    if (chunk === 'Hello') {
```

Comment:
This assumes that you receive the string as a single chunk. Which is probably true nearly all of the time but it isn't guaranteed, it might be split across multiple chunks.

The only thing `socket.setEncoding()` does is ensure the stream won't emit partial multi-byte sequences.

Reply:
That is true and it is not an excuse but there are a lost of tests written like this. I would not add unneeded complexity unless the test turns out to be flaky.

Reply:
Also if the `'data'` handler is not called with a chunk which is not `'Hello'` or `'Bye'`, the assertion in the `else` branch will fail.

Reply:
Right, but our CI is _really_ good at shaking out edge cases and debugging a failing test takes 10x as long as just writing it correctly in the first place.

Reply:
@bnoordhuis does this look better to you:

```js
diff --git a/test/parallel/test-tls-connect-allow-half-open-option.js b/test/parallel/test-tls-connect-allow-half-open-option.js
index 129d1d1ced..1bc7eda1ca 100644
--- a/test/parallel/test-tls-connect-allow-half-open-option.js
+++ b/test/parallel/test-tls-connect-allow-half-open-option.js
@@ -21,6 +21,7 @@ const tls = require('tls');
   assert.strictEqual(socket.allowHalfOpen, false);
 }
 
+const hello = Buffer.from('Hello');
 
 const server = tls.createServer({
   key: fixtures.readKey('agent1-key.pem'),
@@ -28,15 +29,24 @@ const server = tls.createServer({
 }, common.mustCall((socket) => {
   server.close();
 
-  socket.setEncoding('utf8');
-  socket.on('data', common.mustCall((chunk) => {
-    if (chunk === 'Hello') {
-      socket.end(chunk);
+  const chunks = [];
+
+  socket.on('data', (chunk) => {
+    chunks.push(chunk);
+
+    const buf = Buffer.concat(chunks);
+
+    if (buf.equals(hello)) {
+      chunks.length = 0;
+      socket.end(buf);
     } else {
-      assert.strictEqual(chunk, 'Bye');
       assert(!socket.writable);
     }
-  }, 2));
+  });
+
+  socket.on('end', common.mustCall(() => {
+    assert.deepStrictEqual(Buffer.concat(chunks), Buffer.from('Bye'));
+  }));
 }));
 
 server.listen(0, common.mustCall(() => {
@@ -45,19 +55,22 @@ server.listen(0, common.mustCall(() => {
     rejectUnauthorized: false,
     allowHalfOpen: true,
   }, common.mustCall(() => {
-    socket.on('data', common.mustCall((chunk) => {
-      assert.strictEqual(chunk, 'Hello');
-    }));
+    const chunks = [];
+
+    socket.on('data', (chunk) => {
+      chunks.push(chunk);
+    });
+
     socket.on('end', common.mustCall(() => {
+      assert.deepStrictEqual(Buffer.concat(chunks), hello);
+
       setTimeout(() => {
         assert(socket.writable);
         assert(socket.write('Bye'));
         socket.end();
-      }, 50);
+      }, 100);
     }));
 
     socket.write('Hello');
   }));
-
-  socket.setEncoding('utf8');
 }));
```

Reply:
Yep, that works. You can also keep the `socket.setEncoding('utf8')` and concatenate until you see the shibboleth.